### ISSUE TITLE
`where` related gradient bug in jax.nn.softmax and jax.nn.log_softmax (unprotected use of jnp.exp?)

### ISSUE THREAD

**Issue by hrbigelow** [google/jax#17117]
### Description

I see `nan`s in the gradient of `jax.nn.softmax` and `jax.nn.log_softmax`, specifically when I use `where`, and the logits that are filtered out by the `where` condition are large, but not large elsewhere in the logits tensor:

```python
def fn(x, w):
    return jax.sum(jax.nn.softmax(x, axis=0, where=w, initial=0.0))

def lfn(x, w):
    return jnp.sum(jax.nn.log_softmax(x, axis=0, where=w, initial=0.0))

x = jnp.array([1.0, 1.0, 100.0]) 
w = jnp.array([True, True, False])

jax.grad(fn)(x, w)
Array([nan, nan, nan], dtype=float32)

jax.grad(lfn)(x, w)
Array([ 0.,  0., nan], dtype=float32)
```

I believe there is an easy fix, and the bug is due to the fact that `jnp.exp` is being called over the values `x - x_max`, which may overflow since `x_max` is only taken over the values passing the `where` condition.  This is already warned against in the documentation for `jnp.where`:

[https://jax.readthedocs.io/en/latest/faq.html#gradients-contain-nan-where-using-where](https://jax.readthedocs.io/en/latest/faq.html#gradients-contain-nan-where-using-where)

```python
def _softmax(
    x,
    axis: Optional[Union[int, tuple[int, ...]]] = -1,
    where: Optional[Array] = None,
    initial: Optional[Array] = None) -> Array:

# if where is used, x_max may not reflect the global maximum across axis
  x_max = jnp.max(x, axis, where=where, initial=initial, keepdims=True)

# then, here, exp may yield inf
  unnormalized = jnp.exp(x - x_max)
  result = unnormalized / jnp.sum(unnormalized, axis, where=where, keepdims=True)
  if where is not None:
    result = jnp.where(where, result, 0)
  return result
```

Instead, I tried this which provides proper gradients:

```python
def softmax(x, axis, where=None, initial=None):
    x_max = jnp.max(x, axis, where=where, initial=initial, keepdims=True)
    shifted = x - x_max
    if where is not None:
        shifted = jnp.where(where, shifted, -1000.0)
    unnormalized = jnp.exp(shifted)
    result = unnormalized / jnp.sum(unnormalized, axis, where=where, keepdims=True)
    if where is not None:
        result = jnp.where(where, result, 0.0)
```

### What jax/jaxlib version are you using?

jax 0.4.2

### Which accelerator(s) are you using?

GPU

### Additional system info

Ubuntu

**Comment by jakevdp** [comment]
Thanks for following up!
